enum.c

```diff
@@ -851,47 +854,163 @@ ary_inject_op(VALUE ary, VALUE init, VALUE op)
+	ASSUME(op != Qundef);
 	break;
     }
+    *pinit = init;
+    return op;
+}
 
-    if (iter == inject_op_i &&
-        SYMBOL_P(op) &&
-        RB_TYPE_P(obj, T_ARRAY) &&
-        rb_method_basic_definition_p(CLASS_OF(obj), id_each)) {
-        return ary_inject_op(obj, init, op);
-    }
+static VALUE
+rb_enum_inject_call(VALUE obj, VALUE init, VALUE op, rb_block_call_func *iter)
+{
+    struct MEMO *memo;
 
     memo = MEMO_NEW(init, Qnil, op);
     rb_block_call(obj, id_each, 0, 0, iter, (VALUE)memo);
     if (memo->v1 == Qundef) return Qnil;
     return memo->v1;
 }
 
+static VALUE
+accumulate_array(struct MEMO *memo)
+{
+    VALUE result = memo->v2;
+    if (!result) {
+        result = rb_ary_new();
+        MEMO_V2_SET(memo, result);
+        if (memo->v1 != Qundef) rb_ary_push(result, memo->v1);
+    }
+    return result;
+}
+
+static VALUE
+accumulate_i(RB_BLOCK_CALL_FUNC_ARGLIST(i, memop))
+{
+    struct MEMO *memo = MEMO_CAST(memop);
+    VALUE result = accumulate_array(memo);
+    inject_i(i, memop, argc, argv, blockarg);
+    rb_ary_push(result, memo->v1);
+    return Qnil;
+}
+
+static VALUE
+accumulate_op_i(RB_BLOCK_CALL_FUNC_ARGLIST(i, memop))
+{
+    struct MEMO *memo = MEMO_CAST(memop);
+    VALUE result = accumulate_array(memo);
+    inject_op_i(i, memop, argc, argv, blockarg);
+    rb_ary_push(result, memo->v1);
+    return Qnil;
+}
+
+/* v1: initial value/Qundef, v2: result array/Qfalse, u3.value: op/Qundef */
+struct MEMO *
+rb_enum_accumulate_memo_new(int argc, VALUE *argv)
+{
+    VALUE init, op = rb_enum_inject_prepare(argc, argv, &init);
+    return MEMO_NEW(init, Qfalse, op);
+}
+
+VALUE
+rb_enum_accumulate_memo_call(struct MEMO *memo, VALUE proc, VALUE i)
+{
+    if (proc) {
+        VALUE v = memo->v1;
+        if (v != Qundef) {
+            i = rb_proc_call_with_block(proc, 2, (VALUE[2]){v, i}, Qnil);
+        }
+	MEMO_V1_SET(memo, i);
+    }
+    else {
+        inject_op_i(i, (VALUE)memo, 1, &i, Qnil);
+    }
+    return memo->v1;
+}
+
+/*
+ *  call-seq:
+ *     enum.accumulate(initial, sym) -> array
+ *     enum.accumulate(sym)          -> array
+ *     enum.accumulate(initial) { |memo, obj| block }  -> array
+ *     enum.accumulate          { |memo, obj| block }  -> array
+ *
+ *  Produces a collection of cumulative results starts from the <i>initial</i> value by
+ *  processing each element of the input series.
+ *
+ *  If you specify a block, then for each element in <i>enum</i>
+ *  the block is passed an accumulator value (<i>memo</i>) and the element.
+ *  If you specify a symbol instead, then each element in the collection
+ *  will be passed to the named method of <i>memo</i>.
+ *  In either case, the result becomes the new value for <i>memo</i>.
+ *
+ *  If you do not explicitly specify an <i>initial</i> value for <i>memo</i>,
+ *  then the first element of collection is used as the initial value
+ *  of <i>memo</i>.
+ *
+ *  The result collection has a size bigger by one than the initial collection.
```

Comment:
This will be a better description I recon "In the case when an <i>initial</i> value is set the result collection will have a size bigger by one than <i>enum</i> size otherwise the size will be equal to the size of <i>enum</i>."